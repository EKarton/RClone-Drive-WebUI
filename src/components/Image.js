import { useEffect, useState } from 'react';
import useRCloneClient from 'hooks/useRCloneClient';
import { Skeleton } from '@mui/material';
import cx from 'classnames';
import './Image.scss';

export default function Image({ image, width, height, imgClassName, skeletonClassName }) {
  const rCloneClient = useRCloneClient();
  const [imageUrl, setImageUrl] = useState();
  const [error, setError] = useState();

  useEffect(() => {
    const fetchData = async () => {
      try {
        setError(undefined);
        setImageUrl(undefined);

        const response = await rCloneClient.fetchImage(
          image.remote,
          image.folderPath,
          image.fileName
        );

        const imageUrl = URL.createObjectURL(await response.blob());
        setImageUrl(imageUrl);
      } catch (error) {
        setError(error);
      }
    };

    if (!imageUrl && image) {
      fetchData();
    }

    return () => {
      if (imageUrl) {
        URL.revokeObjectURL(imageUrl);
      }
    };
  }, [image, imageUrl, rCloneClient]);

  if (!imageUrl && !error) {
    return (
      <Skeleton
        variant="rectangular"
        className={cx('image', skeletonClassName)}
        width={width}
        height={height}
      />
    );
  }

  if (error) {
    return (
      <div width={width} height={height}>
        Error: {error.message}
      </div>
    );
  }

  return (
    <img
      className={cx('image', imgClassName)}
      src={imageUrl}
      alt={image.fileName}
      loading="lazy"
      width={width}
      height={height}
    />
  );
}
