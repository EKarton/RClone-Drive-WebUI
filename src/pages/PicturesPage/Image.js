import { createRef, useEffect } from "react";

const Image = ({ onResize, ...props }) => {
  const imgRef = createRef();

  useEffect(() => {
    const handleWindowResized = () => {
      onResize(imgRef);
    };

    window.addEventListener("resize", handleWindowResized);
    handleWindowResized();

    return () => {
      window.removeEventListener("resize", handleWindowResized);
    };
  }, [imgRef, onResize]);

  return <img {...props} ref={imgRef} alt={props.alt} />;
};

export default Image;
