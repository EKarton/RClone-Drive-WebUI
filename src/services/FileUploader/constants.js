export const ActionTypes = Object.freeze({
  UPLOAD_FILE: 'UPLOAD_FILE',
  CANCEL_UPLOAD: 'CANCEL_UPLOAD',
});
