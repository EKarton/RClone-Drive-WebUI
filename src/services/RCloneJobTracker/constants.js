export const ActionTypes = Object.freeze({
  TRACK_NEW_JOB: 'TRACK_NEW_JOB',
  UNTRACK_NEW_JOB: 'UNTRACK_NEW_JOB',
});
