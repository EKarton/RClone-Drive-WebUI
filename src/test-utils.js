import { FileViewerProvider, initialState } from 'contexts/FileViewerStore';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router';

const customRender = (
  component,
  { initialFileViewerState = initialState } = {},
  { route = '/' } = {}
) => {
  const history = createMemoryHistory({ initialEntries: [route] });

  return render(
    <FileViewerProvider defaultState={initialFileViewerState}>
      <Router history={history}>{component}</Router>
    </FileViewerProvider>
  );
};

export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { customRender, userEvent };
