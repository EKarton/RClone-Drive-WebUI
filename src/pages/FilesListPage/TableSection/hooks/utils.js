import { JobStatus, JobTypes } from 'utils/constants';

export function getFileNames(data) {
  return new Set(data.filter((file) => !file.IsDir).map((file) => file.Name));
}

export function getFolderNames(data) {
  return new Set(data.filter((file) => file.IsDir).map((file) => file.Name));
}

export function getUploadingJobs(jobs, remote, path) {
  return jobs.filter(
    (job) =>
      job.jobType === JobTypes.UPLOAD_FILE &&
      job.remote === remote &&
      job.dirPath.startsWith(path) &&
      job.status.value !== JobStatus.SUCCESS
  );
}

export function getUploadingFiles(jobs, remote, path) {
  return getUploadingJobs(jobs, remote, path).filter((job) => job.dirPath === path);
}

export function getUploadingFolders(jobs, remote, path) {
  return getUploadingJobs(jobs, remote, path).filter((job) => job.dirPath !== path);
}

export function getMovingJobs(jobs, remote, path) {
  const supportedJobTypes = new Set([
    JobTypes.MOVE_FILE,
    JobTypes.MOVE_FOLDER,
    JobTypes.RENAME_FILE,
    JobTypes.RENAME_FOLDER,
  ]);

  return jobs.filter((job) => {
    const correctJob = supportedJobTypes.has(job.jobType);
    const inProgress = job.status.value === JobStatus.ONGOING;

    if (!correctJob || !inProgress) {
      return false;
    }

    const isMatchSrc = job.src?.remote === remote && job.src?.dirPath === path;
    const isMatchTarget = job.target?.remote === remote && job.target?.dirPath === path;

    return correctJob && inProgress && (isMatchSrc || isMatchTarget);
  });
}
