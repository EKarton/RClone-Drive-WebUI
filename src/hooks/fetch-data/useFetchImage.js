import axios from 'axios';
import { useEffect, useRef, useState } from 'react';
import { StatusTypes } from 'utils/constants';
import useImageFetcher from '../rclone/useImageFetcher';

export default function useFetchImage(image) {
  const cancelSource = useRef(null);
  const imageFetcher = useImageFetcher();

  const [result, setResult] = useState({
    status: StatusTypes.LOADING,
    data: undefined,
    error: undefined,
  });

  useEffect(() => {
    const fetchData = async () => {
      try {
        setResult({ status: StatusTypes.LOADING, data: undefined, error: undefined });

        const cancelToken = cancelSource.current.token;
        const opt = { cancelToken };
        const response = await imageFetcher.getImage(
          image.remote,
          image.dirPath,
          image.fileName,
          opt
        );

        setResult({
          status: StatusTypes.SUCCESS,
          data: URL.createObjectURL(new Blob([response.data])),
          error: undefined,
        });
      } catch (error) {
        if (!axios.isCancel(error)) {
          setResult({ status: StatusTypes.ERROR, data: undefined, error });
        }
      }
    };

    cancelSource.current = axios.CancelToken.source();

    if (!result.data && image) {
      fetchData();
    }

    return () => {
      cancelSource.current.cancel();

      if (result.data) {
        URL.revokeObjectURL(result.data);
      }
    };
  }, [image, imageFetcher, result.data]);

  return result;
}
