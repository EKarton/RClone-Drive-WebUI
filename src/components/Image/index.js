import Skeleton from '@mui/material/Skeleton';
import cx from 'classnames';
import PropTypes from 'prop-types';
import useFetchImage from 'hooks/fetch-data/useFetchImage';
import { StatusTypes } from 'utils/constants';
import './index.scss';

export default function Image({ image, width, height, imgClassName, skeletonClassName }) {
  const { status, data: imageUrl, error } = useFetchImage(image);

  if (status === StatusTypes.LOADING) {
    return (
      <Skeleton
        variant="rectangular"
        className={cx('image', skeletonClassName)}
        width={width}
        height={height}
        data-testid="image-spinner"
      />
    );
  }

  if (status === StatusTypes.ERROR) {
    return (
      <div width={width} height={height} data-testid="image-error">
        Error: {error.message}
      </div>
    );
  }

  return (
    <img
      className={cx('image', imgClassName)}
      src={imageUrl}
      alt={image.fileName}
      loading="lazy"
      width={width}
      height={height}
      data-testid="image-content"
    />
  );
}

Image.propTypes = {
  image: PropTypes.shape({
    remote: PropTypes.string,
    dirPath: PropTypes.string,
    fileName: PropTypes.string,
  }),
  width: PropTypes.number,
  height: PropTypes.number,
  imgClassName: PropTypes.string,
  skeletonClassName: PropTypes.string,
};
