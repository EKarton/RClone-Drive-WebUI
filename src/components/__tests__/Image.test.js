import Image from 'components/Image';
import useFetchImage from 'hooks/rclone/fetch-data/useFetchImage';
import { customRender, waitFor } from 'test-utils/react';
import { StatusTypes } from 'utils/constants';

jest.mock('hooks/rclone/fetch-data/useFetchImage');

describe('Image', () => {
  const imageProps = {
    remote: 'googledrive',
    folderPath: 'Pictures/2020',
    fileName: 'dog.png',
  };

  it('should render spinner when api call is in flight', async () => {
    useFetchImage.mockReturnValue({
      status: StatusTypes.LOADING,
    });

    const component = customRender(
      <Image image={imageProps} width="20px" height="20px" />
    );

    await waitFor(() => {
      expect(component.getByTestId('image-spinner')).toBeInTheDocument();
      expect(component.baseElement).toMatchSnapshot();
    });
  });

  it('should render image when api call finishes', async () => {
    useFetchImage.mockReturnValue({
      status: StatusTypes.SUCCESS,
      data: 'blob://data',
    });

    const component = customRender(
      <Image image={imageProps} width="20px" height="20px" />
    );

    await waitFor(() => {
      expect(component.getByTestId('image-content')).toBeInTheDocument();
      expect(component.baseElement).toMatchSnapshot();
    });
  });

  it('should render error message when api call fails', async () => {
    useFetchImage.mockReturnValue({
      status: StatusTypes.ERROR,
      error: new Error('Image not found'),
    });

    const component = customRender(
      <Image image={imageProps} width="20px" height="20px" />
    );

    await waitFor(() => {
      expect(component.getByTestId('image-error')).toBeInTheDocument();
      expect(component.baseElement).toMatchSnapshot();
    });
  });
});
