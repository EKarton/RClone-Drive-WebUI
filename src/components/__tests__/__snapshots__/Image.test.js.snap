// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image should render error message when api call fails 1`] = `
<body>
  <div>
    <div
      data-testid="image-error"
      height="20px"
      width="20px"
    >
      Error: 
      Image not found
    </div>
  </div>
</body>
`;

exports[`Image should render image when api call finishes 1`] = `
<body>
  <div>
    <img
      alt="dog.png"
      class="image"
      data-testid="image-content"
      height="20px"
      loading="lazy"
      src="blob://data"
      width="20px"
    />
  </div>
</body>
`;

exports[`Image should render spinner when api call is in flight 1`] = `
<body>
  <div>
    <span
      class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse image css-58ftfq-MuiSkeleton-root"
      data-testid="image-spinner"
      style="width: 20px; height: 20px;"
    />
  </div>
</body>
`;
